src/main/java/jcifs/smb/SmbFile.java

```diff
@@ -0,0 +1,2380 @@
+                        RequestParam.NO_RETRY);
+                }
+            }
+            else {
+                if ( ctime != 0 || atime != 0 ) {
+                    throw new SmbUnsupportedOperationException("Cannot set creation or access time without CAP_NT_SMBS");
+                }
+                th.send(
+                    new SmbComSetInformation(th.getConfig(), getUncPath(), attrs, mtime - th.getServerTimeZoneOffset()),
+                    new SmbComSetInformationResponse(th.getConfig()));
+            }
+
+            this.attrExpiration = 0;
+        }
+    }
+
+
+    @Override
+    public void setFileTimes( long createTime, long lastLastModified, long lastLastAccess ) throws SmbException {
+        if ( this.fileLocator.isRootOrShare() ) {
+            throw new SmbException("Invalid operation for workgroups, servers, or shares");
+        }
+
+        try {
+            setPathInformation(0, createTime, lastLastModified, lastLastAccess);
+        }
+        catch ( CIFSException e ) {
+            throw SmbException.wrap(e);
+        }
+    }
+
+
+    @Override
+    public void setCreateTime ( long time ) throws SmbException {
+        if ( this.fileLocator.isRootOrShare() ) {
+            throw new SmbException("Invalid operation for workgroups, servers, or shares");
+        }
+
+        try {
+            setPathInformation(0, time, 0L, 0L);
+        }
+        catch ( CIFSException e ) {
+            throw SmbException.wrap(e);
+        }
+    }
+
+
+    @Override
+    public void setLastModified ( long time ) throws SmbException {
+        if ( this.fileLocator.isRootOrShare() ) {
+            throw new SmbException("Invalid operation for workgroups, servers, or shares");
+        }
+
+        try {
+            setPathInformation(0, 0L, time, 0L);
+        }
+        catch ( CIFSException e ) {
+            throw SmbException.wrap(e);
+        }
+    }
+
+
+    @Override
+    public void setLastAccess ( long time ) throws SmbException {
+        if ( this.fileLocator.isRootOrShare() ) {
+            throw new SmbException("Invalid operation for workgroups, servers, or shares");
+        }
+
+        try {
+            setPathInformation(0, 0L, 0L, time);
+        }
+        catch ( CIFSException e ) {
+            throw SmbException.wrap(e);
+        }
+    }
+
+
+    @Override
+    public int getAttributes () throws SmbException {
+        if ( this.fileLocator.isRootOrShare() ) {
+            return 0;
+        }
+        exists();
+        return this.attributes & ATTR_GET_MASK;
+    }
+
+
+    @Override
+    public void setAttributes ( int attrs ) throws SmbException {
+        if ( this.fileLocator.isRootOrShare() ) {
+            throw new SmbException("Invalid operation for workgroups, servers, or shares");
+        }
+        try {
+            setPathInformation(attrs & ATTR_SET_MASK, 0L, 0L, 0L);
+        }
+        catch ( SmbException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)